lib/3scale/backend/job_fetcher.rb

```diff
@@ -105,23 +59,47 @@ def shutdown
 
       private
 
-      def pop_from_queue
-        begin
-          encoded_job = @redis.blpop(*@queues, timeout: @fetch_timeout)
-        rescue Redis::BaseConnectionError, Errno::ECONNREFUSED, Errno::EPIPE => e
-          raise RedisConnectionError.new(e.message)
-        rescue Redis::CommandError => e
-          # Redis::CommandError from redis-rb can be raised for multiple
-          # reasons, so we need to check the error message to distinguish
-          # connection errors from the rest.
-          if e.message == 'ERR Connection timed out'.freeze
-            raise RedisConnectionError.new(e.message)
-          else
-            raise e
-          end
+      # @param max [Integer] maximum number of records to return
+      #   note: we need to update redis-rb gem so that sync client can also support it
+      # @return [nil, Array<String>] array of strings or `nil` if no results
+      def try_pop_from_queue(max=nil)
```

Comment:
Same as above, I don't see the point on having this method if we are going to end up calling `wait_pop_from_queue` anyway when the queue is empty, and assuming it's not empty, `wait_pop_from_queue` will return immediately as `try_pop_from_queue` would do, so `try_pop_from_queue` seems redundant to me.

Reply:
`try_pop_from_queue` will return multiple results. The idea is that we fetch jobs as fast as possible with `try_..` and if queues are empty, block until something comes up. That is for efficiency. Under high load we will be fetching fast, under low load we don't needlessly perform many queries.

Reply:
> I don't understand this part, under low load this will always call blpop as it did before this PR, it will be the same, right?

`blpop` will be only called if redis list is empty/missing. Under low load I expect us to end-up using `blpop` far more often, but then if load is low, doing a few extra calls shouldn't bother the database or any other component.

I believe the benefits during high load far outweigh the effect during low-load.